lib/AST/Type.cpp

```diff
@@ -976,6 +976,18 @@ bool TypeBase::isAnyObject() {
   return canTy.getExistentialLayout().isAnyObject();
 }
 
+// Distinguish between class-bound types that might be AnyObject vs other
+// class-bound types. Only types that are potentially AnyObject might have a
+// transparent runtime type wrapper like __SwiftValue. This must look through
+// all optional types because dynamic casting sees through them.
+bool TypeBase::isPotentiallyAnyObject() {
+  Type unwrappedTy = lookThroughAllOptionalTypes();
+  if (auto archetype = unwrappedTy->getAs<ArchetypeType>()) {
+    return !archetype->hasRequirements();
+  }
+  return unwrappedTy->isAnyObject();
```

Comment:
Do you want this to be true for Any as well as AnyObject?

Reply:
AnyObject is very different from Any as regards casting behavior, so I suspect Andy wants this to only be true for AnyObject.

Reply:
@slavapestov Good question. If we cared about `Any`, then the API would need to be called `isAnyOrAnyObject()`. I don't think it's possible for an `Any` to be `AnyObject`. That would imply that an existential contains another existential. I noticed the type system code does an Any-or-AnyObject checks in some places, but this API is specifically for potentially `AnyObject`.

The client of this API cares about cases like this. What's important to the client is that only `AnyObject` can hold a `__SwiftValue`. `Any` is never a `__SwiftValue`:

```
class C: Hashable, Equatable, P {
  static func == (lhs: C, rhs: C) -> Bool {
    return true
  }

  func hash(into hasher: inout Hasher) {}
}

let val: AnyHashable = C()
let anyObject = val as AnyObject // metatype == "class __SwiftValue"
let any = val as Any // metatype == "struct AnyHashable"
```